packages/playwright-core/src/server/frames.ts

```diff
@@ -1276,12 +1271,37 @@ export class Frame extends SdkObject {
     taskData: T,
     options: types.TimeoutOptions & types.StrictOptions & { mainWorld?: boolean, querySelectorAll?: boolean, logScale?: boolean, omitAttached?: boolean } = {}): Promise<R> {
 
-    const info = this._page.parseSelector(selector, options);
     const callbackText = body.toString();
-    const data = this._contextData.get(options.mainWorld ? 'main' : info.world)!;
 
     return controller.run(async progress => {
       progress.log(`waiting for selector "${selector}"`);
+
+      while (progress.isRunning()) {
+        progress.log(`waiting for selector "${selector}"`);
+        const { frame, info } = await this._resolveFrame(progress, selector, options);
+        try {
+          return await frame._scheduleRerunnableTaskInFrame(progress, info, callbackText, taskData, options);
+        } catch (e) {
+          if (js.isJavaScriptErrorInEvaluate(e) || isSessionClosedError(e))
+            throw e;
+          continue;
+        }
+      }
+      return undefined as any;
+    }, this._page._timeoutSettings.timeout(options));
+  }
+
+  private async _scheduleRerunnableTaskInFrame<T, R>(
+    progress: Progress,
+    info: SelectorInfo,
+    callbackText: string,
+    taskData: T,
+    options: types.TimeoutOptions & types.StrictOptions & { mainWorld?: boolean, querySelectorAll?: boolean, logScale?: boolean, omitAttached?: boolean }): Promise<R> {
+    if (!progress.isRunning())
+      return undefined as any;
```

Comment:
You can instead call `progress.throwIfAborted()`, to avoid unnecessary errors of "cannot read foo of undefined".